app/views/support/ABTest.scala

```diff
@@ -6,26 +6,51 @@ import java.net.URLEncoder
 import java.nio.charset.StandardCharsets
 
 import com.gu.i18n.CountryGroup
+import com.gu.i18n.CountryGroup._
 import play.api.libs.json._
 import play.api.mvc.{Cookie, Request}
 import play.twirl.api.Html
+import scala.util.Random
 
 import scalaz.NonEmptyList
+import views.html.fragments.giraffe.{contributeAmountButtons, contributeMessage}
 
 sealed trait TestTrait {
-
   type VariantFn
-  case class Variant(variantName: String, variantSlug: String, weight: Double, render: VariantFn) {
+  case class Variant(variantName: String, variantSlug: String, weight: Double, render: VariantFn, countryGroupFilter: Set[CountryGroup] = Set.empty) {
     def testName = name
     def testSlug = slug
+    def matches(countryGroup:CountryGroup):Boolean = countryGroupFilter.isEmpty || countryGroupFilter.contains(countryGroup)
   }
 
   def name: String
   def slug: String
   def variants: NonEmptyList[Variant]
-  val weight: Double = variants.map(_.weight).list.fold(0.0)(_ + _)
-  val cdf : Seq[Double]= variants.map(_.weight).list.foldLeft(Seq[Double]())( (l,p) => l :+ l.lastOption.getOrElse(0.0) + p/weight)
-  val weightedVariants = cdf.zip(variants.list)
+
+  def ValidateVariants = {
+    val countryWithNoVariants = variantsByCountry.collectFirst { case (country, variants) if variants.isEmpty => country.name }
+    val errorMessage = countryWithNoVariants.map(country => s"${getClass.getName}: No variant defined for $country.")
+
+    errorMessage.foreach(m => throw new IllegalStateException(m))
+  }
+
+  val variantsByCountry: Map[CountryGroup, List[Variant]] = {
+
+    def filterVariants(countryGroup: CountryGroup) = variants.list.filter(_.matches(countryGroup))
+
+    CountryGroup.allGroups.map(country => country -> filterVariants(country)).toMap
+    }
+
+  val weightedVariantsByCountry = {
+
+    def addWeights(filteredVariants: List[Variant]) = {
```

Comment:
I like all this stuff! But because it's not that simple, maybe it's worth adding a couple of unit tests for this? Asserting that the cumulative distribution function adds up to 1, that sort of thing?